File: 1472-design-browser-history/1472-design-browser-history.java
class BrowserHistory {
    Integer curr;
    HashMap<Integer, String> history;
    
    public BrowserHistory(String homepage) {
        this.curr = 1;
        history = new HashMap<>();
        history.put(curr, homepage);
    }
    
    public void visit(String url) {
        HashMap<Integer, String> map = this.history;
        int index = curr + 1;
        while(this.history.containsKey(index)){
            this.history.remove(index);
            map.remove(index);
            ++index;
        }
        ++this.curr;
        history.put(curr, url);
        
    }
    
    public String back(int steps) {
        HashMap<Integer, String> map = this.history;
        int max = this.history.size();
        if(steps >= max || curr - steps <= 1){
            this.curr = 1;
        } else{
            this.curr -= steps;
        }
        return this.history.get(curr);
    }
    
    public String forward(int steps) {
        HashMap<Integer, String> map = this.history;
        int current = this.curr;
        int max = this.history.size();
        if(steps + curr >= max){
            curr = max;
        } else{
            curr += steps;
        }
        return this.history.get(curr);

    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */

File: 1472-design-browser-history/1472-design-browser-history.py
from collections import deque
class BrowserHistory:

    def __init__(self, homepage: str):
        self.homepage = homepage
        self.back_page = []
        self.forward_page = deque()
        self.curr = homepage

    def visit(self, url: str) -> None:
        self.back_page.append(self.curr)
        self.forward_page = deque()
        self.curr = url

    def back(self, steps: int) -> str:
        while len(self.back_page) > 0 and steps:
            steps -= 1
            self.forward_page.appendleft(self.curr)
            self.curr = self.back_page.pop()
        return self.curr
    
    def forward(self, steps: int) -> str:
        while len(self.forward_page) > 0 and steps:
            steps -= 1
            self.back_page.append(self.curr)
            self.curr = self.forward_page.popleft()
        return self.curr

# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)